### PR TITLE
Migration/stavetie

### PR DESCRIPTION
**Pull request by rvilarl**
stavetie, tabtie & tabslide!

There are differences in the images due to a bug. Font weight was ignored.

### CODE REVIEW

File: src/stavetie.ts [hunk 1]
```diff
@@ -122,51 +160,66 @@ export class StaveTie extends Element {
...
+    let stave: Stave;
+    const firstNoteStave = this.notes.first_note ? this.notes.first_note.getStave() : undefined;
+    const lastNoteStave = this.notes.last_note ? this.notes.last_note.getStave() : undefined;
+    if (firstNoteStave) {
+      stave = firstNoteStave;
+    } else if (lastNoteStave) {
+      stave = lastNoteStave;
+    } else {
+      throw new Vex.RERR('NoStave', 'Stave required');
+    }
```
Comment (0xfe) [inline src/stavetie.ts:177]: Seems like you can get the same effect with:

```typescript
let stave = this.notes.first_note?.getStave() ?? this.notes.last_note?.getStave();

...
ctx.fillText(..., stave!.getYForTopText() - 1)
...
```

File: src/stavetie.ts [hunk 2]
```diff
@@ -69,26 +110,23 @@ export class StaveTie extends Element {
     }
 
     // Success. Lets grab 'em notes.
-    this.first_note = notes.first_note;
-    this.first_indices = notes.first_indices;
-    this.last_note = notes.last_note;
-    this.last_indices = notes.last_indices;
+    this.notes = notes;
```
Comment (0xfe) [inline src/stavetie.ts:113]: This is okay, do the regression tests show any diffs?

File: src/stavetie.ts [hunk 3]
```diff
@@ -6,9 +6,52 @@
...
+}
 
 export class StaveTie extends Element {
-  constructor(notes, text) {
+  render_options: StaveTieRenderOptions;
+
+  protected text?: string;
+
+  protected font: FontInfo;
+  protected notes: {
```
Comment (0xfe) [inline src/stavetie.ts:38]: Lots of repetition here with this structure, can you make it a type?

File: src/stavetie.ts (hunk 2 above)
Comment (rvilarl) [reply src/stavetie.ts:113]: No diffs, the information was simply duplicated.

File: src/stavetie.ts (hunk 3 above)
Comment (rvilarl) [reply src/stavetie.ts:38]: Done `Notes`

File: src/stavetie.ts [hunk 4]
```diff
@@ -6,9 +6,52 @@
...
+  last_x_shift: number;
+  tie_spacing: number;
+  cp1: number;
+  first_x_shift: number;
+  text_shift_x: number;
+  y_shift: number;
+  font: FontInfo;
+}
+
+export interface StaveTieRenderTieParams {
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/stavetie.ts:24]: done

File: src/stave.ts [hunk 5]
```diff
@@ -341,12 +341,12 @@ export class Stave extends Element {
     return (y - this.y) / spacing - headroom;
   }
 
-  getYForTopText(line: number): number {
+  getYForTopText(line?: number): number {
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/stave.ts:344]: done

File: src/stave.ts [hunk 6]
```diff
@@ -341,12 +341,12 @@ export class Stave extends Element {
...
     const l = line || 0;
     return this.getYForLine(-l - this.options.top_text_position);
   }
 
-  getYForBottomText(line: number): number {
+  getYForBottomText(line?: number): number {
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/stave.ts:349]: done

File: src/stavetie.ts (hunk 1 above)
Comment (rvilarl) [reply src/stavetie.ts:177]: done

Comment (rvilarl) [reply src/stavetie.ts:177]: done

**Comment by rvilarl** [comment]
Ready

**Comment by rvilarl** [comment]
One change was missing, ready

File: src/stavetie.ts [hunk 7]
```diff
@@ -6,9 +6,38 @@
 
 import { Vex } from './vex';
 import { Element } from './element';
+import { FontInfo, RenderTieParams } from './types/common';
+import { Note } from './note';
+import { Stave } from './stave';
```
Comment (0xfe) [inline src/stavetie.ts:11]: Lint error -- `Stave` defined but never used.

File: src/stavetie.ts [hunk 8]
```diff
@@ -6,9 +6,38 @@
...
+
+export interface Notes {
```
Comment (0xfe) [inline src/stavetie.ts:13]: Lets call this `TieNotes`

File: src/stavetie.ts [hunk 9]
```diff
@@ -6,9 +6,38 @@
...
+  first_note: Note;
+  last_note: Note;
+  first_indices: number[];
+  last_indices: number[];
+}
+
+export interface StaveTieRenderOptions {
```
Comment (0xfe) [inline src/stavetie.ts:20]: This can be inline.

File: src/stavetie.ts (hunk 8 above)
Comment (0xfe) [reply src/stavetie.ts:13]: This can also be in common.

File: src/types/common.d.ts [hunk 10]
```diff
@@ -104,6 +104,14 @@ export interface RenderContext {
   measureText(text: string): { width: number };
 }
 
+export interface RenderTieParams {
```
Comment (0xfe) [inline src/types/common.d.ts:107]: On second thought, let's just inline this :-) I prefer the duplication to this very specific interface.

File: src/stavetie.ts (hunk 7 above)
Comment (rvilarl) [reply src/stavetie.ts:11]: done

File: src/stavetie.ts (hunk 8 above)
Comment (rvilarl) [reply src/stavetie.ts:13]: done

File: src/stavetie.ts (hunk 9 above)
Comment (rvilarl) [reply src/stavetie.ts:20]: done

File: src/types/common.d.ts (hunk 10 above)
Comment (rvilarl) [reply src/types/common.d.ts:107]: done :)

**Comment by rvilarl** [comment]
Ready

File: src/tabslide.ts [hunk 11]
```diff
@@ -39,38 +42,44 @@ export class TabSlide extends TabTie {
...
     }
 
-    const ctx = this.context;
+    const ctx = this.checkContext();
     const first_x_px = params.first_x_px;
     const first_ys = params.first_ys;
     const last_x_px = params.last_x_px;
 
-    const direction = this.slide_direction;
+    const direction = params.direction;
```
Comment (0xfe) [inline src/tabslide.ts:76]: I think this should be `this.direction`

File: src/tabtie.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    this.setAttribute('type', 'TabTie');
+
+    this.render_options.cp1 = 9;
+    this.render_options.cp2 = 11;
+    this.render_options.y_shift = 3;
+
+    this.direction = -1; // Tab tie's are always face up.
+
+    this.setNotes(notes);
+  }
```
Comment (0xfe) [inline src/tabtie.ts:41]: Did you intentionally remove `draw`?

File: src/tabslide.ts (hunk 11 above)
Comment (rvilarl) [reply src/tabslide.ts:76]: Done. The result is also fine. However, this means that params.direction is not used, is this ok?

File: src/tabtie.ts (hunk 12 above)
Comment (rvilarl) [reply src/tabtie.ts:41]: Yes, I believe it is identical to the implementation in the parent.

**Comment by rvilarl** [comment]
@0xfe anything pending here?

File: src/stavetie.ts [hunk 13]
```diff
@@ -35,18 +50,19 @@ export class StaveTie extends Element {
...
-      font: { family: 'Arial', size: 10, style: '' },
+      font: { family: 'Arial', size: 10, weight: '' },
     };
 
     this.font = this.render_options.font;
     this.setNotes(notes);
   }
 
-  setFont(font) {
+  setFont(font: FontInfo): this {
```
Comment (0xfe) [inline src/stavetie.ts:60]: It's good that you changed `style` to `weight` but this breaks the public interface, so add a console warning, and also set `this.font.weight = font.style`.

Comment (rvilarl) [reply src/stavetie.ts:60]: I believe that I did not break the interface, I fixed a bug. The value was simply ignored. The images resulting are now different. I would not add console log, to say there was a bug.

Comment (0xfe) [reply src/stavetie.ts:60]: Oh, sorry. I thought `style` was being used. Thanks.

**Comment by 0xfe** [comment]
Looks like there's a conflict in `common.d.ts` now.

**Comment by rvilarl** [comment]
Rebased and conflict solved.
@0xfe if the open conversation is ok, ready to merge.

**Comment by rvilarl** [comment]
@0xfe anything pending here?

File: src/tabslide.ts (hunk 11 above)
Comment (0xfe) [reply src/tabslide.ts]: I'm confused about this... we have both `this.slide_direction` and `params.direction`, so it seems like we can't just ignore `this.direction`. How about `const direction = params.direction ?? this.slide_direction`? (And make `params.direction` optional?)

Comment (0xfe) [reply src/tabslide.ts]: Okay, actually, I dug into the code and see what's going on now -- `slide_direction` is different from `direction`, and this class does not implement `draw`. So your original `this.slide_direction` was correct. Sorry about that, can you revert this line please?

Comment (rvilarl) [reply src/tabslide.ts:76]: done :)

**Comment by rvilarl** [comment]
Ready to merge?

**Comment by 0xfe** [comment]
Great, thanks!

**Pull request merged**